parser_library/src/workspaces/macro_cache.cpp

```diff
@@ -0,0 +1,202 @@
+        if (std::find(cached_instr.begin(), cached_instr.end(), opcode.opcode) != cached_instr.end())
+            result.push_back(
+                { from, opcode.opcode, std::holds_alternative<context::macro_def_ptr>(opcode.opcode_detail) });
+
+        if (std::find(cached_instr.begin(), cached_instr.end(), from) != cached_instr.end())
+            result.push_back(
+                { from, opcode.opcode, std::holds_alternative<context::macro_def_ptr>(opcode.opcode_detail) });
+    }
+
+    // Also macros with the same name as CA instructions may alias the instructions
+    for (const auto& [id, macro] : ctx.macros())
+    {
+        if (!macro)
+            continue;
+        auto opcode = ctx.get_operation_code(id);
+        // Macros for which there is an opsyn in effect are already captured
+        if (auto pval = std::get_if<context::macro_def_ptr>(&opcode.opcode_detail); !pval || *pval != macro)
+            continue;
+
+        // If there is an opsyn, that aliases an instruction to be CA instruction, add it to result
+        if (std::find(cached_instr.begin(), cached_instr.end(), macro->id) != cached_instr.end())
+            result.push_back({ id, id, true });
+    }
+
+    sort_opsyn_state(result);
+
+    return result;
+}
+
+macro_cache_key macro_cache_key::create_from_context(context::hlasm_context& hlasm_ctx, library_data data)
+{
+    return { hlasm_ctx.opencode_file_name(), data, get_opsyn_state(hlasm_ctx) };
+}
+
+void macro_cache_key::sort_opsyn_state(std::vector<cached_opsyn_mnemo>& opsyn_state)
+{
+    std::sort(opsyn_state.begin(), opsyn_state.end(), [](const cached_opsyn_mnemo& lhs, const cached_opsyn_mnemo& rhs) {
+        return std::tie(lhs.from_instr, lhs.to_instr, lhs.is_macro)
+            < std::tie(rhs.from_instr, rhs.to_instr, rhs.is_macro);
+    });
+}
+
+const analyzer* macro_cache::find_cached_analyzer(const macro_cache_key& key) const
+{
+    auto it = cache_.find(key);
+    if (it == cache_.end())
+        return nullptr;
+
+    const auto& cached_data = it->second;
+
+    for (const auto& [fname, cached_version] : cached_data.stamps)
+    {
+        auto file = file_mngr_->find(fname);
+        if (!file)
+            return nullptr; // Reparse needed
+        if (file->get_version() != cached_version)
+            return nullptr;
+    }
+
+    // Version of all dependent files are the same.
+    return cached_data.cached_analyzer.get();
+}
+
+bool macro_cache::load_from_cache(const macro_cache_key& key, const analyzing_context& ctx)
+{
+    if (auto cached_analyzer = find_cached_analyzer(key))
+    {
+        if (key.data.proc_kind == processing::processing_kind::MACRO)
+        {
+            lsp::macro_info_ptr info = cached_analyzer->context().lsp_ctx->get_macro_info(key.data.library_member);
+            if (!info)
+                return true; // The file for which the analyzer is cached does not contain definition of macro
```

Comment:
so is this a case of an invalid file in a macro library?

Reply:
yes, exactly